Automattic/wp-calypso
# Full Site Editing: Theme and Front-end Rendering

## WARNING

This is a prototype! The code is rough!
It's mostly intended to be installed and tested.
Hopefully there are no infinite loops anymore, but you never know.

#### Changes proposed in this Pull Request

### Template Selection

Users can choose the template for Posts and Pages through the new Template sidebar.
It contains a `wp_template_id` meta box with an autocomplete input searching for Templates.

Notes:
- The Template sidebar should be available for all `viewable` CPTs, not just Posts and Pages.
- The `wp_template_id` autocomplete should only search for Templates containing a Content Slot block.
- The meta box should be clearable in order to use the default template.

### Template Hierarchy

Add **barebone and very rough** support for template hierarchy.
Think of it like an abstraction/simplification of the [Core hierarchy](https://developer.wordpress.org/files/2014/10/Screenshot-2019-01-23-00.20.04.png).

Users can assign a Template to a "screen" (for lack of a better term) through the new Template Hierarchy sidebar.
It contains a list of checkbox for all the screens that could use a template.
Currently:
- `index`: the default fallback template for all screens.
- `singular`: the default for `is_singular()` screens.
- All `viewable` post types (e.g. `post`, `page`, `media`): the default for non-singular screens of the given post type (e.g. archives, or posts list page).

### Blank Theme

Add a very minimalistic Blank Theme, which works **very roughly** like the Core template selection.
It has two behaviours, depending if the screen `is_singular` or otherwise.

Singular screens are straightforward: render the first available template; then the Content Slot block inside it will render the post content.
The template selection order for singular screens is:
- Post's own `wp_template_id`.
- `$template_hierarchy->singular`.
- `$template_hierarchy->index`.

Non-singular screens are a bit more complicated: select the first available template; then the Content Slot block inside it will render **the entire posts list**.
The template selection order for non-singular screens is:
- `$template_hierarchy->$post_type`.
- `$template_hierarchy->index`.

Notes:
- This implementation means that Content Slot is responsible of the HTML formatting of the list. This is not ideal imho, but I don't have a better idea right now.
- In the future we should probably give the users the possibility of creating a template for list items (e.g. a template with "slot blocks" for post title, excerpt, comments count, etc.).

#### Screenshots

| Template List | Homepage Template |
| - | - |
| ![Screenshot 2019-05-07 at 18 33 10](https://user-images.githubusercontent.com/2070010/57320792-a3de5200-70f7-11e9-9c85-3b6aa148599d.png) | ![Screenshot 2019-05-07 at 18 33 45](https://user-images.githubusercontent.com/2070010/57320802-a8a30600-70f7-11e9-99cd-bfa47e59c55d.png) |

| Sample Page (Editor) | Sample Page (Rendered) |
| - | - |
| ![Screenshot 2019-05-08 at 17 15 00](https://user-images.githubusercontent.com/2070010/57390684-d8641380-71b4-11e9-8a2f-27d9c4238531.png) | ![Screenshot 2019-05-07 at 18 34 27](https://user-images.githubusercontent.com/2070010/57320826-bd7f9980-70f7-11e9-963a-7f07a9caaf75.png) |

| Template Hierarchy Sidebar | Posts List (Rendered) |
| - | - |
| ![Screenshot 2019-05-10 at 16 59 36](https://user-images.githubusercontent.com/2070010/57540723-40e1fa80-7345-11e9-84b3-a79bc8508305.png) | ![Screenshot 2019-05-10 at 16 59 54](https://user-images.githubusercontent.com/2070010/57540732-45a6ae80-7345-11e9-9c8e-47755e964cbf.png) |

Notes:
- This PR does not contain any CSS, not even obtained from Twenty Nineteen.
- The fake navigation menu is just a column block with three links inside. For some reason, it's not displayed as `flex` in the editor, and so it awkwardly wraps the columns when rendered in preview, but looks just fine when rendered in the front-end.

#### Testing instructions

* Follow the [FSE readme](https://github.com/Automattic/wp-calypso/blob/e759e056360b72576e124989c4174afef78bc1db/apps/full-site-editing/README.md#local-development), and setup both the plugin and the theme.
* In the test WP, enable both the FSE plugin and the Blank theme.
* Create a Template containing some blocks (or Template Parts of other Templates), and a Content Slot block.
* Create or edit a page or a post, open the Template sidebar, and select the template created earlier.
* Save and view the page.
* It should display the whole template, with the post content instead of the Content Slot block.
* Edit the template, open the Template Hierarchy sidebar, and check "Default".
* Open the posts page of the site.
* It should display the whole template, with the posts list (just linked title and content, and without pagination) instead of the Content Slot block.

Fixes #32496
Fixes #32617
